exec/java-exec/src/test/java/org/apache/drill/TestTpchDistributedConcurrent.java

```diff
@@ -0,0 +1,177 @@
+import org.junit.Test;
+import org.junit.rules.TestRule;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.fail;
+
+/*
+ * Note that the real interest here is that the drillbit doesn't become
+ * unstable from running a lot of queries concurrently -- it's not about
+ * any particular order of execution. We ignore the results.
+ */
+// TODO(cwestin) @Ignore
+public class TestTpchDistributedConcurrent extends BaseTestQuery {
+  @Rule public final TestRule TIMEOUT = TestTools.getTimeoutRule(120000); // Longer timeout than usual.
+
+  /*
+   * Valid test names taken from TestTpchDistributed. Fuller path prefixes are
+   * used so that tests may also be taken from other locations -- more variety
+   * is better as far as this test goes.
+   */
+  private final static String queryFile[] = {
+    "queries/tpch/01.sql",
+    "queries/tpch/03.sql",
+    "queries/tpch/04.sql",
+    "queries/tpch/05.sql",
+    "queries/tpch/06.sql",
+    "queries/tpch/07.sql",
+    "queries/tpch/08.sql",
+    "queries/tpch/09.sql",
+    "queries/tpch/10.sql",
+    "queries/tpch/11.sql",
+    "queries/tpch/12.sql",
+    "queries/tpch/13.sql",
+    "queries/tpch/14.sql",
+    // "queries/tpch/15.sql", this creates a view
+    "queries/tpch/16.sql",
+    "queries/tpch/18.sql",
+    "queries/tpch/19_1.sql",
+    "queries/tpch/20.sql",
+  };
+
+  private final static int TOTAL_QUERIES = 115;
+  private final static int CONCURRENT_QUERIES = 15;
+
+  private final static Random random = new Random(0xdeadbeef);
+  private final static String alterSession = "alter session set `planner.slice_target` = 10";
+
+  private int remainingQueries = TOTAL_QUERIES - CONCURRENT_QUERIES;
+  private final Semaphore completionSemaphore = new Semaphore(0);
+  private final Semaphore submissionSemaphore = new Semaphore(0);
+  private final IdentityHashMap<UserResultsListener, Object> listeners = new IdentityHashMap<>();
+
+  private void submitRandomQuery() {
+    final String filename = queryFile[random.nextInt(queryFile.length)];
+    final String query;
+    try {
+      query = QueryTestUtil.normalizeQuery(getFile(filename)).replace(';', ' ');
+    } catch(IOException e) {
+      throw new RuntimeException("Caught exception", e);
+    }
+    final UserResultsListener listener = new ChainingSilentListener(query);
+    client.runQuery(UserBitShared.QueryType.SQL, query, listener);
+    synchronized(listeners) {
+      listeners.put(listener, listener);
+    }
+  }
+
+  private class ChainingSilentListener extends SilentListener {
+    private final String query;
+
+    public ChainingSilentListener(final String query) {
+      this.query = query;
+    }
+
+    @Override
+    public void queryCompleted(QueryState state) {
+      super.queryCompleted(state);
+
+      final Object object;
+      synchronized(listeners) {
+        object = listeners.remove(this);
+      }
+      assertNotNull("listener not found", object);
```

Comment:
I tried it with a hard fail there to see what happens. The test does fail, but by timing out. Is that good enough?